File: HW/HW7/0226-HW7.py
# youre a sysadmin for a big tech company, and you have to estimate the amount of servers you need to handle some long running jobs
# one server can handle only one job at once, one job can only be executed by one server at once. once a server has finished executing its current job, it can be reassigned to a new job.

# given an array of time intervals:
# intervals where intervals[i] = [starti, endi]
# representing the start and end time for a particular job that needs to be executed return the minimum number OverflowErrorservers required to run all
from typing import List
import heapq


def minMeetingRooms(intervals: List[List[int]]) -> int:
    # one server can handle only one job at once
    # one job can only be executed by one server at once
    # once a server has finished executing its current job, it can be reassigned

    intervals.sort()
    servers = 0 # number of servers initialized
    ends = [] # empty list to keep track of end times

    for start, end in intervals:
            # Check if any server is freed up (end time <= current start time)
            while ends and ends[0] <= start:
                heapq.heappop(ends) # popping the earliest end time
            heapq.heappush(ends, end) # push the end time of current job
            servers = max(servers, len(ends))   # update the number of servers
    return servers

testcase1 = [[0,30], [5,10], [15,20]]
print(minMeetingRooms(testcase1))
testcase2 = [[0,1],[1,2],[2,3]]
print(minMeetingRooms(testcase2))

File: HW/HW7/0226-HW7.java
package HW.HW7;
import java.util.Arrays;
import java.util.PriorityQueue;

public class MeetingRooms {
    public int minMeetingRooms(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        PriorityQueue<Integer> ends = new PriorityQueue<>();
        int roomsNeeded = 0;
        for (int[] interval : intervals) {
            int start = interval[0];
            int end = interval[1];
            while (!ends.isEmpty() && ends.peek() <= start) {
                ends.poll();
            }
            ends.offer(end);
            roomsNeeded = Math.max(roomsNeeded, ends.size());
        }
        return roomsNeeded;
    }//minmeetingrooms
    public static void main(String[] args) {
        MeetingRooms meetingRooms = new MeetingRooms();
        
        int[][] testcase1 = {{0, 30}, {5, 10}, {15, 20}};
        System.out.println(meetingRooms.minMeetingRooms(testcase1));

        int[][] testcase2 = {{0, 1}, {1, 2}, {2, 3}};
        System.out.println(meetingRooms.minMeetingRooms(testcase2));
    }//main
}
